File: src/CADL.Extension/Emitter.Csharp/src/type/InputParameter.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { InputConstant } from "./InputConstant";
import { InputOperationParameterKind } from "./InputOperationParameterKind";
import { InputType } from "./InputType";
import { RequestLocation } from "./RequestLocation";

//TODO: Define VirtualParameter for HLC
export interface VirtualParameter {}
export interface InputParameter {
    Name: string;
    NameInRequest: string;
    Description?: string;
    Type: InputType;
    Location: RequestLocation;
    DefaultValue?: InputConstant;
    VirtualParameter?: VirtualParameter; //for HLC, set null for cadl
    GroupedBy?: InputParameter;
    Kind: InputOperationParameterKind;
    IsRequired: boolean;
    IsApiVersion: boolean;
    IsResourceParameter: boolean;
    IsContentType: boolean;
    IsEndpoint: boolean;
    SkipUrlEncoding: boolean;
    Explode: boolean;
    ArraySerializationDelimiter?: string;
    HeaderCollectionPrefix?: string;
}

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputParameter.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;

namespace AutoRest.CSharp.Common.Input;

internal record InputParameter(
    string Name,
    string NameInRequest,
    string? Description,
    InputType Type,
    RequestLocation Location,
    InputConstant? DefaultValue,
    // TODO: This should be removed, tracking it in https://github.com/Azure/autorest.csharp/issues/4779
    InputParameter? GroupedBy,
    InputModelProperty? FlattenedBodyProperty,
    InputOperationParameterKind Kind,
    bool IsRequired,
    bool IsApiVersion,
    bool IsResourceParameter,
    bool IsContentType,
    bool IsEndpoint,
    bool SkipUrlEncoding,
    bool Explode,
    string? ArraySerializationDelimiter,
    string? HeaderCollectionPrefix)
{
    public InputParameter() : this(
        Name: string.Empty,
        NameInRequest: string.Empty,
        Description: null,
        Type: InputPrimitiveType.Unknown,
        Location: RequestLocation.None,
        DefaultValue: null,
        FlattenedBodyProperty: null,
        GroupedBy: null,
        Kind: InputOperationParameterKind.Method,
        IsRequired: false,
        IsApiVersion: false,
        IsResourceParameter: false,
        IsContentType: false,
        IsEndpoint: false,
        SkipUrlEncoding: false,
        Explode: false,
        ArraySerializationDelimiter: null,
        HeaderCollectionPrefix: null)
    { }

    public string Name { get; internal set; } = Name;
    public bool IsRequired { get; internal set; } = IsRequired;
    public InputType Type { get; internal set; } = Type;
    public InputOperationParameterKind Kind { get; internal set; } = Kind;
    public IReadOnlyList<InputDecoratorInfo> Decorators { get; internal set; } = new List<InputDecoratorInfo>();
}
